File: Arrays/P030.py
def brute(nums):
    n=len(nums)
    res=set()
    for i in  range(n-2):
        for j in range(i+1,n-1):
            for k in range(j+1,n):
                if nums[i]+nums[j]+nums[k]==0:
                    res.add(tuple(sorted([nums[i],nums[j],nums[k]])))
    return [list(triplets) for triplets in res]

def optimal(nums):
    n=len(nums)
    nums.sort()
    res=[]
    for i in range(n-2):
        if i>0 and nums[i]==nums[i-1]:
            continue
        left=i+1
        right=n-1
        while (left<right):
            if nums[i]+nums[left]+nums[right]==0:
                res.append([nums[i],nums[left],nums[right]])
                left+=1
                right-=1
                while left<right and nums[left-1]==nums[left]:
                    left+=1
                while left<right and nums[right+1]==nums[right]:
                    right+=1
            elif nums[i]+nums[left]+nums[right]<0:
                left+=1
            else:
                right-=1
    return res

nums=input("Enter array: ").split()
nums=list(map(int,nums))
print("Brute: ",brute(nums))
print("Optimal: ",optimal(nums))

File: Striver_C++_Problems/Arrays/P030.cpp
// P030.cpp - 3Sum: all unique triplets with sum 0

#include <bits/stdc++.h>
using namespace std;

vector<vector<int>> brute(const vector<int>& nums) {
    int n = nums.size();
    set<vector<int>> res_set;
    for (int i = 0; i < n - 2; i++) {
        for (int j = i + 1; j < n - 1; j++) {
            for (int k = j + 1; k < n; k++) {
                if (nums[i] + nums[j] + nums[k] == 0) {
                    vector<int> trip = {nums[i], nums[j], nums[k]};
                    sort(trip.begin(), trip.end());
                    res_set.insert(trip);
                }
            }
        }
    }
    vector<vector<int>> res(res_set.begin(), res_set.end());
    return res;
}

vector<vector<int>> optimal(vector<int> nums) {
    int n = nums.size();
    sort(nums.begin(), nums.end());
    vector<vector<int>> res;

    for (int i = 0; i < n - 2; i++) {
        if (i > 0 && nums[i] == nums[i - 1]) continue;
        int left = i + 1;
        int right = n - 1;
        while (left < right) {
            long long s = (long long)nums[i] + nums[left] + nums[right];
            if (s == 0) {
                res.push_back({nums[i], nums[left], nums[right]});
                left++;
                right--;
                while (left < right && nums[left - 1] == nums[left]) left++;
                while (left < right && nums[right + 1] == nums[right]) right++;
            } else if (s < 0) {
                left++;
            } else {
                right--;
            }
        }
    }
    return res;
}

void printTriplets(const vector<vector<int>>& trips) {
    cout << "[\n";
    for (auto &t : trips) {
        cout << "  [";
        for (size_t i = 0; i < t.size(); i++) {
            cout << t[i];
            if (i + 1 < t.size()) cout << ", ";
        }
        cout << "]\n";
    }
    cout << "]\n";
}

int main() {
    cout << "Enter array: ";
    string line;
    getline(cin, line);
    stringstream ss(line);
    vector<int> nums;
    int x;
    while (ss >> x) nums.push_back(x);

    auto b = brute(nums);
    auto o = optimal(nums);

    cout << "Brute: \n";
    printTriplets(b);
    cout << "Optimal:\n";
    printTriplets(o);
    return 0;
}
